### DISCUSSION TITLE
is nextauth getServerSession() killing nextjs14 cache?

### DISCUSSION THREAD

**Discussion by lfgtavora** [vercel/next.js#59468] [Help]
### Summary

i notice something strange when i use getServerSession from NextAuth with nextjs14 (in main layout), all the fetchs from pages and components requests are losing deafult cache, when i remove this method every things works fine, i tried to use useSession() too, and works fine but i would like to avoid because only work on client components and shows a "layout shifting". I watched few videos on yt but nobody seems notice that.

without getServerSession i tried fetch simple timer api :
```
dateTime: '2023-12-09T18:27:51.1726568', 
dateTime: '2023-12-09T18:27:51.1726568', 
dateTime: '2023-12-09T18:27:51.1726568', 
dateTime: '2023-12-09T18:27:51.1726568', 
dateTime: '2023-12-09T18:27:51.1726568',
```
with getServerSession i tried fetch simple timer api :
```
  dateTime: '2023-12-09T19:11:25.1488771',
  dateTime: '2023-12-09T19:11:27.306346',
  dateTime: '2023-12-09T19:12:19.1603039',
  dateTime: '2023-12-09T19:12:33.7091574',
```
----
here is complete scenario:

i using getServerSession in layout.tsx(server component), to fullfill my navbar(client component) with user informations. Then in the page.tsx i make 2 calls(1 for article and another for comments).

await fetch(`http:localhost:3000/api/posts/${username}/${slug}`)
and comments.tsx (server component) are inside a a <suspense>

await fetch(`http:localhost:3000/api/comments?postId=${postId}&page=${page}`, { next: { tags: ['comments'] } })
when i invoke getServerSession in layout every refresh on page calls my apis again (i put a log there). But if i remove getServerSession from layout and use useSession() directly in navbar.tsx(client component) the fetch uses cache and not call my api again, just after a revalidate.

here are the header list logged in my api

```
HeadersList {
  cookies: null,
  [Symbol(headers map)]: Map(11) {
    'accept' => { name: 'accept', value: '*/*' },
    'accept-encoding' => { name: 'accept-encoding', value: 'gzip, deflate' },
    'accept-language' => { name: 'accept-language', value: '*' },
    'connection' => { name: 'connection', value: 'keep-alive' },
    'host' => { name: 'host', value: 'localhost:3000' },
    'sec-fetch-mode' => { name: 'sec-fetch-mode', value: 'cors' },
    'user-agent' => { name: 'user-agent', value: 'undici' },
    'x-forwarded-for' => { name: 'x-forwarded-for', value: '::1' },
    'x-forwarded-host' => { name: 'x-forwarded-host', value: 'localhost:3000' },
    'x-forwarded-port' => { name: 'x-forwarded-port', value: '3000' },
    'x-forwarded-proto' => { name: 'x-forwarded-proto', value: 'http' }
  },
  [Symbol(headers map sorted)]: [
    [ 'accept', '*/*' ],
    [ 'accept-encoding', 'gzip, deflate' ],
    [ 'accept-language', '*' ],
    [ 'connection', 'keep-alive' ],
    [ 'host', 'localhost:3000' ],
    [ 'sec-fetch-mode', 'cors' ],
    [ 'user-agent', 'undici' ],
    [ 'x-forwarded-for', '::1' ],
    [ 'x-forwarded-host', 'localhost:3000' ],
    [ 'x-forwarded-port', '3000' ],
    [ 'x-forwarded-proto', 'http' ]
  ]
}
```

### Additional information

```js
System:
    OS: Windows 10 10.0.19045
    CPU: (6) x64 Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz
    Memory: 7.97 GB / 15.94 GB
  Binaries:
    Node: 18.17.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 9.6.7 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.3636.0), Chromium (120.0.2210.61)
    Internet Explorer: 11.0.19041.3636
  npmPackages:
    @auth/core: ^0.18.4 => 0.18.4
    @auth/prisma-adapter: ^1.0.9 => 1.0.9
    next: ^14.0.4 => 14.0.4
    next-auth: ^5.0.0-beta.4 => 5.0.0-beta.4
    react: ^18.2.0 => 18.2.0
```

**Comment by jfaithedu** [comment]
When you use getServerSession, it checks the requests cookies / headers to authenticate the session. This check will force your pages to become dynamic and hence adopt a "no-store" caching strategy.

When a page is marked as dynamic, the default caching behavior changes. Dynamic pages are typically not cached in the same manner as static pages because their content can change based on different user sessions or other variables. This is why you're observing a different caching behavior when getServerSession is used versus when it is not.

As for the "layout shifting" issue with useSession(), this is likely because useSession() is a client-side hook and it retrieves session information after the initial render.


[From the nextjs docs](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching)

> 
> fetch requests are not cached if:
> 
> The cache: 'no-store' is added to fetch requests.
> The revalidate: 0 option is added to individual fetch requests.
> The fetch request is inside a Router Handler that uses the POST method.
> **The fetch request comes after the usage of headers or cookies.**
> The const dynamic = 'force-dynamic' route segment option is used.
> The fetchCache route segment option is configured to skip cache by default.
> **The fetch request uses Authorization or Cookie headers and there's an uncached request above it in the component tree.**
>
